File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/python/evm/smart-accounts/get_or_create_smart_account.py
# Usage: uv run python evm/smart-accounts/get_or_create_smart_account.py

import asyncio

from cdp import CdpClient
from dotenv import load_dotenv

load_dotenv()


async def main():
    async with CdpClient() as cdp:
        # Create an owner account
        owner = await cdp.evm.create_account()
        print("Created owner account:", owner.address)

        # Create a smart account with the owner
        name = "MySmartAccount"
        account = await cdp.evm.get_or_create_smart_account(name=name, owner=owner)
        print("Smart Account Address: ", account.address)

        # Try to get the same smart account again - should return the existing one
        account2 = await cdp.evm.get_or_create_smart_account(name=name, owner=owner)
        print("Second Smart Account Address: ", account2.address)
        print("Are accounts equal? ", account.address == account2.address)

asyncio.run(main())

File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/typescript/evm/smart-accounts/getOrCreateSmartAccount.ts
// Usage: pnpm tsx evm/smart-accounts/getOrCreateSmartAccount.ts

import { CdpClient } from "@coinbase/cdp-sdk";
import "dotenv/config";

const cdp = new CdpClient();

// First create an owner account
const owner = await cdp.evm.createAccount();
console.log("Created owner account:", owner.address);

// Get or create a smart account with the owner
// Note: Each owner can only have one smart account
const account = await cdp.evm.getOrCreateSmartAccount({ name: "MyAccount", owner });
console.log("EVM Smart Account Address:", account.address);

// Subsequent calls to getOrCreateSmartAccount with the same owner will return the existing account
const sameAccount = await cdp.evm.getOrCreateSmartAccount({ name: "MyAccount", owner });
console.log("Retrieved same account:", sameAccount.address);
console.log("Are accounts equal?", account.address === sameAccount.address); // Will be true

// To create multiple smart accounts, you need different owners
const anotherOwner = await cdp.evm.createAccount();
console.log("\nCreated another owner account:", anotherOwner.address);

const differentAccount = await cdp.evm.getOrCreateSmartAccount({ name: "DifferentAccount", owner: anotherOwner });
console.log("Different EVM Smart Account Address:", differentAccount.address);
